libfeed/feed-knowledge-app-news-card-store.c

```diff
@@ -0,0 +1,107 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-discovery-feed is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-discovery-feed is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-discovery-feed.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "feed-knowledge-app-news-card-store.h"
+#include "feed-knowledge-app-card-store.h"
+#include "feed-sizes.h"
+
+struct _EosDiscoveryFeedKnowledgeAppNewsCardStore
+{
+  GObject parent_instance;
+};
+
+static void base_card_store_iface_init (EosDiscoveryFeedBaseCardStoreInterface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (EosDiscoveryFeedKnowledgeAppNewsCardStore,
+                         eos_discovery_feed_knowledge_app_news_card_store,
+                         EOS_DISCOVERY_FEED_TYPE_KNOWLEDGE_APP_CARD_STORE,
+                         G_IMPLEMENT_INTERFACE (EOS_DISCOVERY_FEED_TYPE_BASE_CARD_STORE,
+                                                base_card_store_iface_init))
+
+enum {
+  PROP_0,
+  PROP_TYPE,
+  NPROPS
+};
+
+static GParamSpec *eos_discovery_feed_knowledge_app_news_card_store_props [NPROPS] = { NULL, };
+
+static void
+eos_discovery_feed_knowledge_app_news_card_store_get_property (GObject    *object,
+                                                               guint       prop_id,
+                                                               GValue     *value,
+                                                               GParamSpec *pspec)
+{
+  EosDiscoveryFeedKnowledgeAppNewsCardStore *store = EOS_DISCOVERY_FEED_KNOWLEDGE_APP_NEWS_CARD_STORE (object);
+
+  switch (prop_id)
+    {
+    case PROP_TYPE:
+      g_value_set_uint (value, EOS_DISCOVERY_FEED_CARD_STORE_TYPE_NEWS_CARD);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
+    }
+}
+
+static void
+eos_discovery_feed_knowledge_app_news_card_store_init (EosDiscoveryFeedKnowledgeAppNewsCardStore *store)
+{
+}
+
+static void
+base_card_store_iface_init (EosDiscoveryFeedBaseCardStoreInterface *iface)
+{
+}
+
+static void
+eos_discovery_feed_knowledge_app_news_card_store_class_init (EosDiscoveryFeedKnowledgeAppNewsCardStoreClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->get_property = eos_discovery_feed_knowledge_app_news_card_store_get_property;
+
+  g_object_class_override_property (object_class,
```

Comment:
Does this work without installing the new property?

Reply:
As I understand it, `g_object_class_override_property` sets the *implementation* on an existing property installed on an interface. You just need to set a prop ID for this class.

So it appears to work without also installing it on the class.

Reply:
What confused me is that all other classes call `g_object_class_install_properties()` with `PROP_TYPE` as an argument; I realize now that they pass `PROP_TYPE` instead of `N_PROPS` only to exclude the last property, which is `PROP_TYPE`.
If it works in both cases, I would simply call `g_object_class_install_properties()` with the full array, both here and in the other classes.

Reply:
Would that actually work? I suppose maybe I could use `g_param_spec_override` ....

Reply:
Okay, simply installing a `GParamSpecOverride` on its own here does not work:

```
(process:1257): GLib-GObject-CRITICAL **: g_param_spec_override: assertion 'G_IS_PARAM_SPEC (overridden)' failed

(process:1257): GLib-GObject-CRITICAL **: validate_pspec_to_install: assertion 'G_IS_PARAM_SPEC (pspec)' failed

(process:1257): GLib-GObject-CRITICAL **: Object class EosDiscoveryFeedKnowledgeAppCardStore doesn't implement property 'type' from interface 'EosDiscoveryFeedBaseCardStore'

(process:1257): GLib-GObject-CRITICAL **: g_param_spec_override: assertion 'G_IS_PARAM_SPEC (overridden)' failed

(process:1257): GLib-GObject-CRITICAL **: validate_pspec_to_install: assertion 'G_IS_PARAM_SPEC (pspec)' failed

(process:1257): GLib-GObject-CRITICAL **: Object class EosDiscoveryFeedKnowledgeAppArtworkCardStore doesn't implement property 'type' from interface 'EosDiscoveryFeedBaseCardStore'

(process:1257): GLib-GObject-CRITICAL **: g_param_spec_override: assertion 'G_IS_PARAM_SPEC (overridden)' failed

(process:1257): GLib-GObject-CRITICAL **: validate_pspec_to_install: assertion 'G_IS_PARAM_SPEC (pspec)' failed

(process:1257): GLib-GObject-CRITICAL **: Object class EosDiscoveryFeedKnowledgeAppNewsCardStore doesn't implement property 'type' from interface 'EosDiscoveryFeedBaseCardStore'

(process:1257): GLib-GObject-CRITICAL **: g_param_spec_override: assertion 'G_IS_PARAM_SPEC (overridden)' failed
```

It looks like `g_object_class_override_property` does a little bit more work in terms of finding the right class to override the property on (https://gitlab.gnome.org/GNOME/glib/blob/master/gobject/gobject.c#L841). Given that this is the case, and I don't want to have to respecify the defaults every time I define the property, I can't do it all in one go with `g_object_class_install_properties` as suggested.